### PR TITLE
[feature](struct-type) support struct_element function

### PR DESCRIPTION
**Pull request by xy720**
# Proposed changes

This commit support a function allows return a field column in struct column.
Since the function can return any type, this commit also supports `ANY_STRUCT_TYPE` 
and `ANY_ELEMENT_TYPE`. 

#16547 

example:

```
mysql> select struct_element(named_struct('f1', 1, 'f2', 'a'), 'f2');
+--------------------------------------------------------+
| struct_element(named_struct('f1', 1, 'f2', 'a'), 'f2') |
+--------------------------------------------------------+
| a                                                      |
+--------------------------------------------------------+
1 row in set (0.03 sec)

mysql> select struct_element(named_struct('f1', 1, 'f2', 'a'), 1);
+-----------------------------------------------------+
| struct_element(named_struct('f1', 1, 'f2', 'a'), 1) |
+-----------------------------------------------------+
|                                                   1 |
+-----------------------------------------------------+
1 row in set (0.02 sec)

mysql> select struct_col, struct_element(struct_col, 'f1') from test_struct;
+-------------------------------------------------+-------------------------------------+
| struct_col                                      | struct_element(`struct_col `, 'f1') |
+-------------------------------------------------+-------------------------------------+
| {1, 2, 3, 4, 5}                                 |                                   1 |
| {1, 1000, 10000000, 100000000000, 100000000000} |                                   1 |
| {5, 4, 3, 2, 1}                                 |                                   5 |
| NULL                                            |                                NULL |
| {1, NULL, 3, NULL, 5}                           |                                   1 |
+-------------------------------------------------+-------------------------------------+
9 rows in set (0.01 sec)
```

## Problem summary

Describe your changes.

## Checklist(Required)

* [ ] Does it affect the original behavior
* [x] Has unit tests been added
* [x] Has document been added or modified
* [ ] Does it need to update dependencies
* [ ] Is this PR support rollback (If NO, please explain WHY)

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by xy720** [comment]
run buildall

**Comment by xy720** [comment]
run buildall

**Comment by xy720** [comment]
run buildall

**Comment by xy720** [comment]
run buildall

**Comment by xy720** [comment]
run buildall

**Comment by xy720** [comment]
run buildall

**Comment by xy720** [comment]
run buildall

**Comment by xy720** [comment]
run buildall

**Comment by xy720** [comment]
run buildall

**Comment by xy720** [comment]
run buildall

**Comment by xy720** [comment]
run buildall

**Comment by xy720** [comment]
run buildall

**Comment by xy720** [comment]
run buildall

**Comment by xy720** [comment]
run beut

**Comment by eldenmoon** [comment]
It seems that `struct_element` will read the whole struct and discard other none related columns，will we implement a function that only read the target column.This could save lots of scan io.

**Review by eldenmoon — APPROVED** [review]
LGTM

File: fe/fe-common/src/main/java/org/apache/doris/catalog/AnyStructType.java [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+// specific language governing permissions and limitations
+// under the License.
+
+package org.apache.doris.catalog;
+
+/**
+ *  Describes a AnyStructType type, used for SQL function return type,
+ *   NOT used for table column type.
+ */
+public class AnyStructType extends AnyType {
```
Comment (amorynan) [inline fe/fe-common/src/main/java/org/apache/doris/catalog/AnyStructType.java:24]: is this describe whole struct type and nested element type ? why not use AnyType?

**Comment by xy720** [comment]
> It seems that `struct_element` will read the whole struct and discard other none related columns，will we implement a function that only read the target column.This could save lots of scan io.

That's a very good idea! How to scan target field is important. But I think what we need is not only a function.
What we should do is that we prune sub fields in struct/map type while generating query plan in Fe.
And then pass the pruned fields into column_reader.new_iterator() function in be, so that we read only the target field in Block, which will be passed to function.
This is very important when we support nested complex type.

File: fe/fe-common/src/main/java/org/apache/doris/catalog/AnyStructType.java (hunk 1 above)
Comment (xy720) [reply fe/fe-common/src/main/java/org/apache/doris/catalog/AnyStructType.java:24]: We need specific types to describe the return field type and return struct type.
So it is the ANY_ELEMENT_TYPE and ANY_STRUCT_TYPE.
If there is only ANY_TYPE, We cannot distinguish between return field type and return struct type.

**Review by cambyzju — APPROVED** [review]
LGTM

**Pull request merged**